services/configstore/types/types.go

```diff
@@ -309,6 +309,8 @@ type Project struct {
 	WebhookSecret string `json:"webhook_secret,omitempty"`
 
 	PassVarsToForkedPR bool `json:"pass_vars_to_forked_pr,omitempty"`
+
+	DefaultBranch string `json:"default_branch,omitempty"`
```

Comment:
What should we do with existing projects that wil have an empty `DefaultBranch`?

Reply:
It's not about the type. Without a pointer it'll be an empty string, with the pointer it'll be nil so in both cases we can assume that (with an empty string or nil) that it isn't populated (since default branch with an empty name can't exist).

I was asking how the handle current repositories where the configstore project object has default_branch not populated. Should we just return it empty value until an user does a refresh? How the ui (when using DefaultBranch) should handle an empty value in such case?

Reply:
@sgotti if I add a loop (in the Gateway or Configstore?) to synch all repository info (with configuration)?
Or you like better, I can check if DefaultBranch is empty when the user call api that return projects, I can refresh repository info with default branch?

Reply:
We are adding new data previously not saved in a project object and we CANNOT provide a default for it (since we can't know it it's master, main or anything else).
For new projects this isn't an issue since the project creation will fail if we can't fetch the repository info from the remote source.
For existing projects we have to update them fetching the required data from the remote source.
But the fetching may fail for many reasons. What should we do when we can't fetch the defaultBranch?
* Can we provide it as empty to the client (and make this part of the api contract)?
* Should we mark the project as "invalid"/"out of sync" and communicate this to the client?

Since this field, at least for the moment, isn't used internally we can return it as empty. The client, will handle such case.

About on how to update them we have different options.

* Asynchronously (with rate limiting) and automatically try to update all projects (your first proposed option). But we should handle the case were updating fails and we must not retry it forever, so we are in a case were we have to count and save the number of failures and mark the project as out of sync and a manual refresh by the user should be required. This is going to be complicated. Perhaps useful in the future...

* Update when trying to fetch a repository info in the gateway. This will happen in many points and we're going to add write logic to a readonly api. And also in such case fetching repository info errors should be handled.

* Leave this to the user (by calling the refresh api) (with the assumption that we can provide it as empty). The client could provide an hint to the user that the project should be refreshed.

I'm for the last option, at least for the moment.

Reply:
Thanks @sgotti. So the user can use the new api /refreshremoterepo for refresh, I leave default_branch empty if we have not this information, for many reasons.
I will implement the tests after gitea will be upgraded.